pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -512,45 +562,184 @@ func (b *BinaryOperation) mergeOneSide(data []types.InstantVectorSeriesData, sou
+	// We also don't know if the output will be exclusively floats or histograms, so we'll use the same size slice for both.
+	// We only assign the slices once we see the associated point type so it shouldn't be common that we allocate both.
 	//
 	// FIXME: this is not safe to do for one-to-many, many-to-one or many-to-many matching, as we may need the input series for later output series.
-	if len(left.Floats) < len(right.Floats) {
-		output = left.Floats[:0]
-		defer b.Pool.PutFPointSlice(right.Floats)
+	var leftPutFPointSlice, leftPutHPointSlice, rightPutFPointSlice, rightPutHPointSlice bool
+	leftPoints := len(left.Floats) + len(left.Histograms)
+	rightPoints := len(right.Floats) + len(right.Histograms)
+	leftSmaller := leftPoints < rightPoints
+	var maxPoints int
+	if leftSmaller {
+		maxPoints = leftPoints
 	} else {
-		output = right.Floats[:0]
-		defer b.Pool.PutFPointSlice(left.Floats)
+		maxPoints = rightPoints
+	}
+
+	getFSlice := func() error {
+		if maxPoints <= len(left.Floats) && (leftSmaller || maxPoints > len(right.Floats)) {
+			// Reuse the left slice if we fit and either:
+			// 1. The left side is smaller and therefore more optimal, or
+			// 2. We can't fit into the (smaller) right side.
+			fPoints = left.Floats[:0]
+			rightPutFPointSlice = true
+		} else if maxPoints <= len(right.Floats) {
+			// Reuse the right slice if we fit, and we have already determined that we can't fit into a smaller left side
+			fPoints = right.Floats[:0]
+			leftPutHPointSlice = true
+		} else {
+			// We can't fit in either left or right side, so create a new slice
+			var err error
+			if fPoints, err = b.Pool.GetFPointSlice(maxPoints); err != nil {
+				return err
+			}
+			leftPutHPointSlice = true
+			rightPutFPointSlice = true
+		}
+		return nil
+	}
+
+	getHSlice := func() error {
+		if maxPoints <= len(left.Histograms) && (leftSmaller || maxPoints > len(right.Histograms)) {
+			// Reuse the left slice if we fit and either:
+			// 1. The left side is smaller and therefore more optimal, or
+			// 2. We can't fit into the (smaller) right side.
+			hPoints = left.Histograms[:0]
+			rightPutHPointSlice = true
+		} else if maxPoints <= len(right.Histograms) {
+			// Reuse the right slice if we fit, and we have already determined that we can't fit into a smaller left side
+			hPoints = right.Histograms[:0]
+			leftPutHPointSlice = true
+		} else {
+			// We can't fit in either left or right side, so create a new slice
+			var err error
+			if hPoints, err = b.Pool.GetHPointSlice(maxPoints); err != nil {
+				return err
+			}
+			leftPutHPointSlice = true
+			rightPutHPointSlice = true
+		}
+		return nil
 	}
 
-	nextRightIndex := 0
+	// InstantVectorSeriesDataIterator returns either the next float or histogram
+	leftIterator := types.NewInstantVectorSeriesDataIterator(&left)
+	rightIterator := types.NewInstantVectorSeriesDataIterator(&right)
```

Comment:
Rather than allocating two new iterators for every call to `computeResult`, we could reuse the same two iterators for the life of the operator and reset them on each call to `computeResult`.